src/main/util.ts

```diff
@@ -247,74 +264,101 @@ const getVideoState = (storageDir: unknown) => {
-    const newestVideoPath = getNewestVideo(storageDir);
-    const newestVideoName = path.basename(newestVideoPath, '.mp4');
-    fs.writeFileSync(storageDir + newestVideoName + ".json", jsonString);
+    const metadataFileName = getMetadataFileForVideo(file);
+
+    return await fspromise.writeFile(metadataFileName, jsonString);
 }    
 
 /**
- * runSizeMonitor, maxStorage in GB.
+ * Return information about a video needed for various parts of the application
  */
-const runSizeMonitor = (storageDir: any, maxStorageGB: any) => {
-    console.debug("Running size monitor");  
-    const maxStorageBytes = maxStorageGB * Math.pow(1024, 3);
-    let totalSize = 0;
-    const files = fs.readdirSync(storageDir);
+const getVideoInfo = (videoPath: string): VideoInfo => {
+    const fstats = fs.statSync(videoPath);
+    const mtime = fstats.mtime.getTime();
+    const size = fstats.size;
 
-    for (const file of files) {
-        totalSize += fs.statSync(storageDir + file).size;
-    }
+    return { name: videoPath, size, mtime };
+};
 
-    if (totalSize > maxStorageBytes) { 
-        deleteOldestVideo(storageDir);
-        runSizeMonitor(storageDir, maxStorageBytes);
-    } 
-}   
+/**
+ * Asynchronously find and return a list of video files in the given directory,
+ * sorted by modification time (newest to oldest)
+ */
+const getSortedVideos = async (storageDir: string): Promise<VideoInfo[]> => {
+    const files = await globPromise(path.join(storageDir, "*.mp4"));
+    return files
+        .map(getVideoInfo)
+        .sort((A: VideoInfo, B: VideoInfo) => B.mtime - A.mtime);
+};
 
 /**
- * Delete the oldest video, unprotected video.
+ * Asynchronously delete the oldest, unprotected videos to ensure we don't store
+ * more material than the user has allowed us.
  */
-const deleteOldestVideo = (storageDir: any) => {
-    const sortedVideos = glob.sync(storageDir + "*.mp4")
-        .map((name: any) => ({name, mtime: fs.statSync(name).mtime}))
-        .sort((A: any, B: any) => B.mtime - A.mtime)
+const runSizeMonitor = async (storageDir: string, maxStorageGB: number): Promise<void> => {
+    const maxStorageBytes = maxStorageGB * Math.pow(1024, 3);
+
+    let files = await getSortedVideos(storageDir);
+    console.debug(`[Size Monitor] Running (max size = ${maxStorageGB} GB)`);
+
+    files = files.map((file: any) => {
+        try {
+            const metadataFileName = getMetadataFileForVideo(file.name);
+            const data = fs.readFileSync(metadataFileName);
+            const metadata = JSON.parse(data.toString());
+            return { ...file, metadata, };
+        } catch (e) {
+            console.log(`Unable read and/or parse JSON from metadata file: ${file.name}`);
+        }
+    });
 
-    let videoForDeletion = sortedVideos.pop();
-    let deleteNotAllowed = isVideoProtected(videoForDeletion.name);
+    // Filter files with metadata
+    files = files.filter((file: any) => file && file.hasOwnProperty('metadata'));
```

Comment:
Do we need `file &&` here? Maybe we do as a null check?

Also another comment - 

This will let videos without metadata build up. Maybe we should just delete them? Not sure what the best behaviour here is. I think I'm leaning towards delete anything without a metadata file.

The app won't display a video without metadata, the counter argument is that it could still be salvaged manually. I doubt many users are going to do that though. 

I've definitely already seen logs from users who have lots of leaked video files without metadata, it would be nice to protect against that.

Reply:
We could use `file?.hasOwnProperty()` yes.

I think it'd be a good idea to simply remove those that don't have metadata. The videos are primarily intended to be viewed and used i the app, so it makes sense to clean up properly after us.

Besides, the filter for no metadata was simply made as a guard against missing metadata which would cause the protected test to fail, but they we could remove the metadata check and simply have the protected check be:

```typescript
!file.hasOwnProperty('metadata') || !Boolean(file.metadata.protected)
```

If it doesn't have metadata, it will be considered for deletion.
If it _has_ metadata, but isn't protected it will also be considered for deletion.

Let me know what you think and I'll fix it whichever way you like the best.

Reply:
Yeah - lets do it!